pkg/controller/elasticsearch/stackmon/volume.go

```diff
@@ -0,0 +1,81 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package stackmon
+
+import (
+	"fmt"
+
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/certificates"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/volume"
+)
+
+const (
+	metricbeatConfigVolumeName   = "metricbeat-config"
+	metricbeatConfigDirMountPath = "/etc/metricbeat-config"
+
+	filebeatConfigVolumeName   = "filebeat-config"
+	filebeatConfigDirMountPath = "/etc/filebeat-config"
+
+	monitoringMetricsSourceEsCaCertVolumeName       = "es-monitoring-metrics-source-certs"
+	monitoringMetricsTargetEsCaCertVolumeNameFormat = "es-monitoring-metrics-target-certs-%d"
+	monitoringLogsTargetEsCaCertVolumeNameFormat    = "es-monitoring-logs-target-certs-%d"
+)
+
+func metricbeatConfigSecretName(es esv1.Elasticsearch) string {
+	return esv1.ESNamer.Suffix(es.Name, metricbeatConfigVolumeName)
+}
+
+func metricbeatConfigSecretVolume(es esv1.Elasticsearch) volume.SecretVolume {
+	return volume.NewSecretVolumeWithMountPath(
+		metricbeatConfigSecretName(es),
+		metricbeatConfigVolumeName,
+		metricbeatConfigDirMountPath,
+	)
+}
+
+func filebeatConfigSecretName(es esv1.Elasticsearch) string {
+	return esv1.ESNamer.Suffix(es.Name, filebeatConfigVolumeName)
+}
+
+func filebeatConfigSecretVolume(es esv1.Elasticsearch) volume.SecretVolume {
+	return volume.NewSecretVolumeWithMountPath(
+		filebeatConfigSecretName(es),
+		filebeatConfigVolumeName,
+		filebeatConfigDirMountPath,
+	)
+}
+
+func monitoringMetricsSourceCaCertSecretVolume(es esv1.Elasticsearch) volume.SecretVolume {
+	return volume.NewSecretVolumeWithMountPath(
+		certificates.PublicCertsSecretName(esv1.ESNamer, es.Name),
+		monitoringMetricsSourceEsCaCertVolumeName,
+		monitoringMetricsSourceEsCaCertMountPath,
+	)
+}
+
+func monitoringMetricsTargetCaCertSecretVolumes(es esv1.Elasticsearch) []volume.VolumeLike {
+	volumes := make([]volume.VolumeLike, 0)
+	for i, assoc := range es.GetMonitoringMetricsAssociation() {
+		volumes = append(volumes, volume.NewSecretVolumeWithMountPath(
+			assoc.AssociationConf().GetCASecretName(),
+			fmt.Sprintf(monitoringMetricsTargetEsCaCertVolumeNameFormat, i),
+			fmt.Sprintf(monitoringMetricsTargetEsCaCertMountPathFormat, i),
+		))
+	}
+	return volumes
+}
+
+func monitoringLogsTargetCaCertSecretVolumes(es esv1.Elasticsearch) []volume.VolumeLike {
+	volumes := make([]volume.VolumeLike, 0)
+	for i, assoc := range es.GetMonitoringLogsAssociation() {
+		volumes = append(volumes, volume.NewSecretVolumeWithMountPath(
+			assoc.AssociationConf().GetCASecretName(),
+			fmt.Sprintf(monitoringLogsTargetEsCaCertVolumeNameFormat, i),
+			fmt.Sprintf(monitoringLogsTargetEsCaCertMountPathFormat, i),
```

Comment:
I wonder if we should adopt a similar approach here to what we do in Agent and not use the order dependent indices as part of the mount path. 

I _think_ it does not matter in this first stage where we only support one reference per type but I wonder if this could become a problem in the future. 

In Agent we use namespace/name instead.

Reply:
Good point.
This implies to generate the `output` section in the beat configuration.
```

Reply:
Good point. I will put this aside as it's not trivial.
I think then it implies to generate the ouput section in the beat config. (I probably should have done this from the start.)

Because currently it references the indices:
```
output.elasticsearch:
  hosts: ['${ES_0_TARGET_URL}']
  username: ${ES_0_TARGET_USERNAME}
  password: ${ES_0_TARGET_PASSWORD}
  ssl.certificate_authorities: ["/mnt/es/0/monitoring/logs/target/ca.crt"]
```
Note: the env var indirection instead of the raw value is useful to consume the password through the secret as env var.

Reply:
I think I'd be +1 with trying to use explicit namespace & names rather than ordinals.
It means the config file needs to be templatized, I guess? Maybe not that hard?